### PR TITLE
Followuser

### PR DESCRIPTION
**Pull request by tmquang1999**

### CODE REVIEW

File: app/controllers/relationships_controller.rb [hunk 1]
```diff
@@ -0,0 +1,21 @@
+class RelationshipsController < ApplicationController
+  before_action :logged_in_user
+  
+  def create
+    @user = User.find(params[:followed_id])
```
Comment (chintk) [inline app/controllers/relationships_controller.rb]: remove()

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -45,6 +45,20 @@ def destroy
     redirect_to users_url
   end
 
+  def following
+    @title = "Following"
```
Comment (chintk) [inline app/controllers/users_controller.rb]: i18n

File: app/controllers/users_controller.rb [hunk 3]
```diff
@@ -45,6 +45,20 @@ def destroy
...
+    @user  = User.find(params[:id])
```
Comment (chintk) [inline app/controllers/users_controller.rb]: remove()

File: app/models/user.rb [hunk 4]
```diff
@@ -39,7 +39,8 @@ def forget
   end
   
   def feed
-    Book.where("user_id = ?", id)
+    Book.where("user_id IN (:following_ids) OR user_id = :user_id",
```
Comment (chintk) [inline app/models/user.rb]: scope

File: app/models/user.rb [hunk 5]
```diff
@@ -39,7 +39,8 @@ def forget
...
+                    following_ids: following_ids, user_id: id)
   end
 
   def authenticated?(attribute, token)
```
Comment (chintk) [inline app/models/user.rb]: remove()

File: app/views/users/_unfollow.html.erb [hunk 6]
```diff
@@ -0,0 +1,5 @@
+<%= form_for(current_user.active_relationships.find_by(followed_id: @user.id),
+             html: { method: :delete },
```
Comment (chintk) [inline app/views/users/_unfollow.html.erb]: indent

File: app/views/users/_follow_form.html.erb [hunk 7]
```diff
@@ -0,0 +1,9 @@
+<% unless current_user?(@user) %>
+  <div id="follow_form">
+  <% if current_user.following?(@user) %>
+    <%= render 'unfollow' %>
+  <% else %>
+    <%= render 'follow' %>
+  <% end %>
+  </div>
+<% end %>
```
Comment (chintk) [inline app/views/users/_follow_form.html.erb]: end line

File: app/views/shared/_stats.html.erb [hunk 8]
```diff
@@ -0,0 +1,15 @@
+<% @user ||= current_user %>
+<div class="stats">
+  <a href="<%= following_user_path(@user) %>">
+    <strong id="following" class="stat">
+      <%= @user.following.count %>
+    </strong>
+    <%=t(".following")%>
```
Comment (chintk) [inline app/views/shared/_stats.html.erb]: remove(), add space

File: app/views/users/show_follow.html.erb [hunk 9]
```diff
@@ -0,0 +1,30 @@
+<% provide(:title, @title) %>
+<div class="row">
+  <aside class="col-md-4">
+    <section class="user_info">
+      <%= gravatar_for @user %>
+      <h1><%= @user.name %></h1>
+      <span><%= link_to "view my profile", @user %></span>
+      <span><b>Books:</b> <%= @user.books.count %></span>
```
Comment (chintk) [inline app/views/users/show_follow.html.erb]: i18n

File: app/views/users/show_follow.html.erb [hunk 10]
```diff
@@ -0,0 +1,30 @@
+<% provide(:title, @title) %>
+<div class="row">
+  <aside class="col-md-4">
+    <section class="user_info">
+      <%= gravatar_for @user %>
+      <h1><%= @user.name %></h1>
+      <span><%= link_to "view my profile", @user %></span>
```
Comment (chintk) [inline app/views/users/show_follow.html.erb]: i18n

File: app/views/users/show_follow.html.erb [hunk 11]
```diff
@@ -0,0 +1,30 @@
+<% provide(:title, @title) %>
```
Comment (chintk) [inline app/views/users/show_follow.html.erb]: remove()

File: app/controllers/relationships_controller.rb [hunk 12]
```diff
@@ -0,0 +1,21 @@
+class RelationshipsController < ApplicationController
+  before_action :logged_in_user
+  
+  def create
+    @user = User.find params[:followed_id]
```
Comment (manhcuongdtbk) [inline app/controllers/relationships_controller.rb]: use find_by

File: app/controllers/relationships_controller.rb [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+    current_user.follow @user
+    respond_to do |format|
+      format.html { redirect_to @user }
+      format.js
+    end
+  end
+  
+  def destroy
+    @user = Relationship.find(params[:id]).followed
```
Comment (manhcuongdtbk) [inline app/controllers/relationships_controller.rb]: use find_by

File: app/controllers/users_controller.rb [hunk 14]
```diff
@@ -45,6 +45,20 @@ def destroy
     redirect_to users_url
   end
 
+  def following
+    @title = t(".following")
```
Comment (manhcuongdtbk) [inline app/controllers/users_controller.rb]: i18n ?

File: app/controllers/users_controller.rb [hunk 15]
```diff
@@ -45,6 +45,20 @@ def destroy
...
+    @user  = User.find params[:id]
```
Comment (manhcuongdtbk) [inline app/controllers/users_controller.rb]: use find_by

File: app/controllers/users_controller.rb [hunk 16]
```diff
@@ -45,6 +45,20 @@ def destroy
...
+    @users = @user.following.paginate(page: params[:page])
+    render 'show_follow'
+  end
+
+  def followers
+    @title = t(".followers")
```
Comment (manhcuongdtbk) [inline app/controllers/users_controller.rb]: i18n ?

File: app/controllers/users_controller.rb [hunk 17]
```diff
@@ -45,6 +45,20 @@ def destroy
     redirect_to users_url
   end
 
+  def following
+    @title = t(".following")
+    @user  = User.find params[:id]
+    @users = @user.following.paginate(page: params[:page])
```
Comment (manhcuongdtbk) [inline app/controllers/users_controller.rb]: remove ()

File: app/controllers/users_controller.rb [hunk 18]
```diff
@@ -45,6 +45,20 @@ def destroy
...
+    render 'show_follow'
+  end
+
+  def followers
+    @title = t(".followers")
+    @user  = User.find params[:id]
```
Comment (manhcuongdtbk) [inline app/controllers/users_controller.rb]: use find_by

File: app/controllers/users_controller.rb [hunk 19]
```diff
@@ -45,6 +45,20 @@ def destroy
...
+    @users = @user.followers.paginate(page: params[:page])
```
Comment (manhcuongdtbk) [inline app/controllers/users_controller.rb]: remove ()

File: app/views/shared/_stats.html.erb [hunk 20]
```diff
@@ -0,0 +1,15 @@
...
+  </a>
+  <a href="<%= followers_user_path(@user) %>">
+    <strong id="followers" class="stat">
+      <%= @user.followers.count %>
+    </strong>
+    <%=t ".followers"%>
```
Comment (manhcuongdtbk) [inline app/views/shared/_stats.html.erb]: i18n ?

File: app/controllers/users_controller.rb [hunk 21]
```diff
@@ -45,6 +45,20 @@ def destroy
     redirect_to users_url
   end
 
+  def following
+    @title = t ".following"
+    @user  = User.find_by id: params[:id]
+    @users = @user.following.paginate page: params[:page]
+    render 'show_follow'
```
Comment (tmquang1999) [inline app/controllers/users_controller.rb]: ""

File: app/controllers/users_controller.rb [hunk 22]
```diff
@@ -45,6 +45,20 @@ def destroy
...
+  end
+
+  def followers
+    @title = t ".followers"
+    @user  = User.find_by id: params[:id]
+    @users = @user.followers.paginate page: params[:page]
+    render 'show_follow'
```
Comment (tmquang1999) [inline app/controllers/users_controller.rb]: ""

File: app/controllers/users_controller.rb [hunk 23]
```diff
@@ -45,6 +45,20 @@ def destroy
...
+    @title = t ".following"
+    @user  = User.find_by id: params[:id]
+    @users = @user.following.paginate page: params[:page]
+    render 'show_follow'
+  end
+
+  def followers
+    @title = t ".followers"
+    @user  = User.find_by id: params[:id]
+    @users = @user.followers.paginate page: params[:page]
```
Comment (tmquang1999) [inline app/controllers/users_controller.rb]: followers

File: app/controllers/users_controller.rb [hunk 24]
```diff
@@ -45,6 +45,20 @@ def destroy
     redirect_to users_url
   end
 
+  def following
+    @title = t ".following"
```
Comment (tmquang1999) [inline app/controllers/users_controller.rb]: ?

File: app/models/book.rb [hunk 25]
```diff
@@ -14,6 +14,8 @@ class Book < ApplicationRecord
   validate :picture_size
 
   scope :recently_published, -> { order(created_at: :desc).limit 6 }
+  scope :find_feed, ->(following_ids, id) { where("user_id IN (:following_ids) OR user_id = :user_id",
```
Comment (chintk) [inline app/models/book.rb]: where user_id: user_ids
user_ids = following_ids << user_id

File: app/models/user.rb [hunk 26]
```diff
@@ -19,6 +19,9 @@ class User < ApplicationRecord
   has_secure_password
   validates :password, presence: true, length: { minimum: 6 }, allow_nil: true
 
+  scope :find_feed, -> { where("user_id IN (:following_ids) OR user_id = :user_id",
```
Comment (chintk) [inline app/models/user.rb]: remove

File: app/views/shared/_stats.html.erb [hunk 27]
```diff
@@ -0,0 +1,15 @@
+<% @user ||= current_user %>
+<div class="stats">
+  <a href="<%= following_user_path(@user) %>">
```
Comment (chintk) [inline app/views/shared/_stats.html.erb]: remove()

File: app/views/users/_follow.html.erb [hunk 28]
```diff
@@ -0,0 +1,4 @@
+<%= form_for(current_user.active_relationships.build, remote: true) do |f| %>
```
Comment (chintk) [inline app/views/users/_follow.html.erb]: remove()

File: app/controllers/users_controller.rb [hunk 29]
```diff
@@ -45,6 +45,20 @@ def destroy
     redirect_to users_url
   end
 
+  def following
```
Comment (chintk) [inline app/controllers/users_controller.rb]: restful

**Pull request merged**